### PR TITLE
[CONTINT-3684] allow communication with dca for sidecar and customising sidecar container image

### PR DESCRIPTION
**Pull request by adel121**
### What does this PR do?

This PR allows adding users have more fine-grained customisation on he sidecar injection feature by allowing:
- disabling/enabling communication between injected sidecar and the cluster-agent (enabled and configured by default)
- setting custom container_registry, image_name or image_tag to set a specific image for the agent sidecar

### Motivation

Allow users to further customise injected sidecar.

### Describe how to test/QA your changes

Same QA as this PR, but:
- set image configs for sidecar image by setting the following env vars on the cluster agent and make sure the sidecar image is constructed based on these env vars:
   - DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_CONTAINER_REGISTRY
   - DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_IMAGE_NAME
   - DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_IMAGE_TAG
- ensure agent sidecar can communicate with the cluster agent
- ensure you can disable communication with the cluster agent

You can make use of the following setup:

- helm installation:
```
datadog:
  apiKeyExistingSecret: datadog-secret
  appKeyExistingSecret: datadog-secret
  kubelet:
    tlsVerify: false

agents:
  enabled: false

clusterAgent:
  enabled: true
  admissionController:
    enabled: true
  env:
    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_ENABLED
      value: "true"
    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_PROVIDER
      value: "fargate"
    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_CONTAINER_REGISTRY
      value: "my-registry"
    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_IMAGE_NAME
      value: "my-image"
    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_IMAGE_TAG
      value: "my-tag"

    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_SELECTORS
      value: '[{"ObjectSelector": {"MatchLabels": {"injectSidecarPodLabel": "true"}}, "NamespaceSelector": {"MatchLabels": {"injectSidecarNs": "true"}}}]'
    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_PROFILES
      value: |
        [{
        "env": [
            {"name": "ENV_VAR_1", "value": "value1"}
        ],
        "resources": {
            "limits": {
                "cpu": "1",
                "memory": "512Mi"
            },
            "requests": {
                "cpu": "0.5",
                "memory": "256Mi"
            }
        }}]
```

- Create a namespace that matches the selector:
```
apiVersion: v1
kind: Namespace
metadata:
  name: ns-for-qa
  labels:
    injectSidecarNs: "true"
```

Create the following service account, clusterrole and clusterrolebinding:
```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: datadog-agent
rules:
  - apiGroups:
      - ""
    resources:
      - nodes
      - namespaces
      - endpoints
    verbs:
      - get
      - list
  - apiGroups:
      - ""
    resources:
      - nodes/metrics
      - nodes/spec
      - nodes/stats
      - nodes/proxy
      - nodes/pods
      - nodes/healthz
    verbs:
      - get
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: datadog-agent
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: datadog-agent
subjects:
  - kind: ServiceAccount
    name: datadog-agent
    namespace: default
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: datadog-agent
  namespace: default
```
Add the generated cluster agent authentication token to the `datadog-secret` secret:
- Get the token first:
```
kubectl get secret datadog-agent-cluster-agent -o yaml 

....
apiVersion: v1
data:
  token: TlhCVVNEdUpOaDlSM3FQQ2d0cnBpVWowYVBDZXM4NEQ=
kind: Secret
....
```

- Add it in datadog-secret:
```
kubectl edit secret datadog-secret

# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: v1
data:
  api-key: ****
  token: ****
  app-key: ****
```

Copy the datadog secret from the default namespace to the `ns-for-qa` namespace:

```
kubectl get secret datadog-secret -n default -o yaml | sed -e 's/namespace: default/namespace:  ns-for-qa/' | kubectl apply -f -
```

Create a pod matching the selectors:
```
apiVersion: v1
kind: Pod
metadata:
  name: sample-pod
  namespace: ns-for-qa
  labels:
    admission.datadoghq.com/enabled: "false"
    injectSidecarPodLabel: "true"
spec:
  containers:
    - name: nginx
      image: nginx:1.14.2
      ports:
        - containerPort: 80
```

- Ensure the agent sidecar is injected with the correct image and cluster agent env vars:
```
kubectl get pod sample-pod -n ns-for-qa -o yaml

....
- env:
    - name: DD_API_KEY
      valueFrom:
        secretKeyRef:
          key: api-key
          name: datadog-secret
    - name: DD_SITE
      value: datadoghq.com
    - name: DD_CLUSTER_NAME
    - name: DD_KUBERNETES_KUBELET_NODENAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: spec.nodeName
    - name: DD_CLUSTER_AGENT_ENABLED
      value: "true"
    - name: DD_CLUSTER_AGENT_AUTH_TOKEN
      valueFrom:
        secretKeyRef:
          key: token
          name: datadog-agent-cluster-agent
    - name: DD_CLUSTER_AGENT_URL
      value: https://datadog-agent-cluster-agent.default.svc.cluster.local:5005
    - name: DD_ORCHESTRATOR_EXPLORER_ENABLED
      value: "true"
    - name: DD_EKS_FARGATE
      value: "true"
    - name: ENV_VAR_1
      value: value1
    image: datadog/agent:latest
    imagePullPolicy: IfNotPresent
    name: datadog-agent-injected
    resources:
      limits:
        cpu: "1"
        memory: 512Mi
      requests:
        cpu: 500m
        memory: 256Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-84c9r
      readOnly: true
....
```

Finally ensure the agent sidecar can connect successfully to the cluster agent:
```
kubectl exec sample-pod -c datadog-agent-injected -n ns-for-qa -- agent status

....
=====================
Datadog Cluster Agent
=====================

  - Datadog Cluster Agent endpoint detected: https://datadog-agent-cluster-agent.default.svc.cluster.local:5005/
  Successfully connected to the Datadog Cluster Agent.
  - Running: 7.52.0-devel+git.648.ff59345.commit.ff59345856
....
```

### CODE REVIEW

**Review by levan-m — APPROVED** [review]

File: releasenotes-dca/notes/enable-dca-in-agent-sidecar-80a2a12b482c1a62.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+features:
+  - |
+    Enable `cluster agent` in injected agent sidecar container by default.
+    Set default configuration for connection with the `cluster agent`.
```
Comment (maycmlee) [inline releasenotes-dca/notes/enable-dca-in-agent-sidecar-80a2a12b482c1a62.yaml:5]:
```suggestion
    Enable `cluster agent` in injected Agent sidecar container by default.
    Set the default configuration for connecting with the `cluster agent`.
```

**Review by maycmlee — COMMENTED** [review]
Left a couple of suggestions, but approving.

**Review by zhuminyi — APPROVED** [review]

**Review by maycmlee — APPROVED** [review]

**Review by sgnn7 — APPROVED** [review]

**Pull request merged**